### PR TITLE
Gradual decommit in wks

### PR DESCRIPTION
**Pull request by PeterSolMS**
This addresses a performance regression with regions which was caused by decommitting memory more agressively than with segments.

The fix is to essentially replicate the logic used with segments - we use the time elapsed since the last GC to compute the size we should decommit in this GC. The goal is to ramp down gradually rather than suddenly, in case the application uses more memory again in the future.

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -12845,8 +12847,23 @@ void gc_heap::distribute_free_regions()
         }
     }
 #else //MULTIPLE_HEAPS
-    while (decommit_step())
+    // we want to limit the amount of decommit we do per time to indirectly
```
Comment (mangod9) [inline src/coreclr/gc/gc.cpp:12850]: is similar logic already implemented for the svr case?

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -12845,8 +12847,23 @@ void gc_heap::distribute_free_regions()
...
+
+    decommit_step (decommit_step_milliseconds);
+
+    // transfer any remaining regions on the decommit list back to the free list
+    for (int kind = basic_free_region; kind < count_free_region_kinds; kind++)
     {
+        if (global_regions_to_decommit[kind].get_num_free_regions() != 0)
+        {
+            gc_heap* hp = pGenGCHeap;
+            hp->free_regions[kind].transfer_regions (&global_regions_to_decommit[kind]);
```
Comment (cshung) [inline src/coreclr/gc/gc.cpp:12865]: I suspect we might need to do something with respect to commit accounting here. Looks like we are moving some committed bytes from one bucket to another.

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:12850]: Yes, in SVR case gc thread #0 wakes up every 100 ms and call decommit_step().

File: src/coreclr/gc/gc.cpp (hunk 2 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:12865]: Interesting - distribute_free_regions right now doesn't seem to do anything concerning commit accounting, as far as I can tell. It sounds like this is an omission we should repair?

Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:12865]: @cshung wouldn't that already be covered by the account in `decommit_step` which actually does the decommit?

**Comment by Maoni0** [comment]
it would be good to take this opportunity to consolidate the decommit we do for regions for WKS. right now we do call `decommit_heap_segment_pages` in `decommit_ephemeral_segment_pages` which can decommit but only for `ephemeral_heap_segment` which is just one of the gen0 regions. and then we can decommit again in `distribute_free_regions`. I think we should only do the decommit for regions for WKS in one place (ie, in `distribute_free_regions`). what do you think? let's talk more in our meeting today.

File: src/coreclr/gc/gc.cpp (hunk 2 above)
Comment (cshung) [reply src/coreclr/gc/gc.cpp:12865]: Thanks, @Maoni0, I didn't notice @PeterSolMS already replied a week ago.

I read the code again and I think you are right, we have moved the committed bytes from the object heap to free through `return_free_region` and discounted from free during `virtual_decommit` caused by `decommit_step`.

The change here is just moving the committed bytes from the per heap free lists (i.e. the free bucket) to the global free list (i.e. again, the free bucket), so there is no need to account for that.

**Comment by PeterSolMS** [comment]
The logic concerning regions in `decommit_ephemeral_segment_pages` is specifically intended to decommit the tails of regions where they exceed the budget. This mostly concerns gen 1 regions where gen 1 budget is small, but may also concern gen 0 if we have a lot of pinning. The logic was added with PR#66008.

I have looked into consolidating the logic in `distribute_free_regions`, but found that having that function deal with parts of regions as well as whole regions would likely make things messier and harder to understand.

So my vote would be to keep this as is.

**Comment by Maoni0** [comment]
@PeterSolMS I meant this actual decommit for WKS GC in `decommit_ephemeral_segment_pages` -

```cplusplus
#ifndef MULTIPLE_HEAPS
    // we want to limit the amount of decommit we do per time to indirectly
    // limit the amount of time spent in recommit and page faults
    size_t ephemeral_elapsed = (size_t)((dd_time_clock (dd0) - gc_last_ephemeral_decommit_time) / 1000);
    gc_last_ephemeral_decommit_time = dd_time_clock (dd0);

    // this is the amount we were planning to decommit
    ptrdiff_t decommit_size = heap_segment_committed (ephemeral_heap_segment) - decommit_target;

    // we do a max of DECOMMIT_SIZE_PER_MILLISECOND per millisecond of elapsed time since the last GC
    // we limit the elapsed time to 10 seconds to avoid spending too much time decommitting
    ptrdiff_t max_decommit_size = min (ephemeral_elapsed, (10*1000)) * DECOMMIT_SIZE_PER_MILLISECOND;
    decommit_size = min (decommit_size, max_decommit_size);

    slack_space = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment) - decommit_size;
    decommit_heap_segment_pages (ephemeral_heap_segment, slack_space);
#endif // !MULTIPLE_HEAPS
```
not what you added at the top of this method that sets the decommit target for regions in SVR GC (that doesn't actually do any decommit). the `decommit_heap_segment_pages` we call for WKS GC makes sense for segments but not for regions because for regions `ephemeral_heap_segment` is just that very first region in gen0. we are already doing decommit for regions in WKS GC in `distribute_free_regions` so there's no need to do this in 2 places, especially when the behavior in `decommit_ephemeral_segment_pages` is questionable.

**Review by Maoni0 — APPROVED** [review]

**Pull request merged**